### DISCUSSION TITLE
Creating a bootloader for a custom board

### DISCUSSION THREAD

**Discussion by kostoskistefan** [chipKIT32/chipKIT-core#472] [Q&A] [answered]
I recently found about chipKIT and I love the project, great job! 

I'm using a PIC32MX170F256D on a custom PCB, but I don't see the PIC listed in any of the supported boards in the repository. I can't seem to find documentation on how to create a bootloader for the MCU with chipKIT. I went through all files in the repository [PIC32-avrdude-bootloader](https://github.com/chipKIT32/PIC32-avrdude-bootloader), but I can't find an already existing bootloader, nor any documentation on how to make my own, so my last resort was asking on this discussion section of Github.

Can I get any information on how to approach building an avrdude bootloader with chipKIT?

Btw, I using a PICKIT3 to program the device, and MPLAB X IDE to write the software.

Thanks to everyone in advance!

**Comment by majenkotech** [comment, marked as answer]
The lack of documentation is basically because creating a custom bootloader is a slightly tricky process and writing a howto would be a long and involved thing that none of us like the idea of tackling :)

The MX170, while supported, isn't actively used by any official boards, so we just haven't built a bootloader for it yet. The process, though is:

1. Create a configuration for the board (config bits and bootloader configuration settings) in configs (and link it from the main `BoardConfig.h` file)
2. If you're using MPLAB-X then:
    1. Copy a similar configuration (an MX270 one is best)
    2. Tweak it to use your new settings and make sure you have the right linker script and are including the right files
    3. Build
3. Or if you're using the Makefile
    1. Copy a similar configuration block in the Makefile
    2. Tweak it to use your new settings and make sure you have the right linker script and are including the right files
    3. Run `make <your config name>`
 
For the MX170 you'll be using the `MX17-27-USB-boot-linkerscript.ld` linker script and you won't want any of the USB stuff, so just `main.o` should be the only file you need to link.

And then of course you need to craft a board configuration if you want to use the Arduino IDE / arduino-cli / UECIDE / PlatformIO etc, and that's another entire kettle of fish...

**Reply by majenkotech** [reply]
Also if you're using MPLAB-X you need the paid for compiler to get the -mips16 option. Using the Makefile you can easily point to the pic32-tools compiler which works.

**Reply by kostoskistefan** [reply]
Thank you so much for taking the time to answer! I really appreciate the help. 
I'll try my best to link and compile everything according to your instructions.

**Reply by majenkotech** [reply]
Creating the config bits settings is best done in MPLAB-X using the Configuration view then "generate source code". Then it's a simple copy-and-paste for that block of settings.

**Comment by JacobChrist** [comment]
Stefan,

What's on your board? It doesn't looks like the chip you are using has the Ethernet module so your probably good, but I know there are some bootloader issues that prevent Ethernet from working that may or may have not gotten pushed to the public repos.

Also, I'm assuming that you found the bootloader MPLAB X repo's.

Jacob

**Reply by kostoskistefan** [reply]
Greetings Jacob,

My board doesn't use any Ethernet or USB support, so that won't be a problem. I'm mostly interested in the ADC, SPI and I2C functionalities (as well as the obvious GPIO capability) because I'm having some trouble with MPLAB, and also I already have a code written for AVR and ARM boards which I'm hoping will be much easier to adapt for the PIC once I'm running the avrdude bootloader. 

About the repo, I'm assuming it's [this one](https://github.com/chipKIT32/PIC32-avrdude-bootloader/tree/master/bootloaders/chipKIT-Bootloaders.X), correct?

Stefan

**Reply by kostoskistefan** [reply]
After hitting a few bumps here and there, I've successfully managed to compile the bootloader. Using the Makefile and pic32-tools was much easier than using MPLAB X (which generated tons of errors for some reason, and that reason is probably me not knowing enough about how everything works in the IDE). Anyhow, I have a few questions now:

1. To upload the bootloader to the PIC32, do I just add it to MPLAB IPE as any other hex file? Does MPLAB IPE know it's a bootloader file (through a magic number in the hex I suppose)?
2. After uploading the bootloader, can I compile a simple test program in Arduino IDE/PlatformIO and upload the hex file through MPLAB IPE?
3. If everything works, when I get a little bit more free time (hopefully around mid December), is it okay if I create either a written or a video guide on creating custom bootloaders using your platform?

**Reply by JacobChrist** [reply]
I just looked and it seems as if there are three bootloader projects.  The one you found seems to be the latest and it seems to be the one I use to build bootloaders.  It's probably been over two years since I've built a bootloader.

1. Yes, use MPLAB IPE or similar tool to flash the MCU.  If you pic the correct device the address will be defined in the HEX file for flashing the chip.  I think I normally burn the chip right in MPLAB IDE when doing development and only go to the IPE when I'm building a batch of boards (10+).
2. I don't know.  Once the bootloader is installed you can upload through the Arduino IDE using the chipKIT project, but you will need to create a variant for your board.  This is done updating several files in this folder: 
 - ...\Arduino15\packages\chipKIT\hardware\pic32\2.x.x
There may be some errors in my brief outline on how to do this but here is the gist:
  - Add you board to boards.txt
  - Create a variant in the variants folder by duplicating one of the existing boards closest to your board.
  - Update the Board_Data.c and Board_Defs.h files to match your board.
  - Restart the IDE
  - Rinse and repeat as needed until your board is working.
  - I think @majenkotech is most up on this process as he has done some of the most recent maintenance on this.  The last time I was heavily working on these file was nearly 10 years ago... Where dose the time go.
  - If you need help, I'm willing to hop on discord or zoom and try to walk you though / refigure out what is needed to be done.
  - If you get this working, the next step would be to get these rolled into the chipKIT project.  This can either be done in a generic way so that it would work for anyone trying to use the chip you are using or more specifically calling out your board.  I would recommend the latter only if you plan on selling your board to the general public so that it helps users identify your specific board.

3. You have my permission, not that you need it!

A final note, I think I have a license for the XC32 that I maybe able to dust off and build the bootloader for you if you are having issues.  But I'm not sure if I can even get it to work because its been so long since I've build a bootloader.

Jacob

**Reply by majenkotech** [reply]
> To upload the bootloader to the PIC32, do I just add it to MPLAB IPE as any other hex file? Does MPLAB IPE know it's a bootloader file (through a magic number in the hex I suppose)?

Yep. Just program it as you would any .HEX file.

> After uploading the bootloader, can I compile a simple test program in Arduino IDE/PlatformIO and upload the hex file through MPLAB IPE?

Yes, although you need to create a proper board profile. That consists of editing the boards.txt file for the chipkit platform to create your own board definition, and create a suitable variant (start by copying an existing one that's close to what you want) with your pinout and such defined in it.  The chipKIT variants are quite complex with many many things in them to configure...

> If everything works, when I get a little bit more free time (hopefully around mid December), is it okay if I create either a written or a video guide on creating custom bootloaders using your platform?

It's not only okay, it's actively encouraged. Let us know where it is and we'll feature it on the chipKIT website :)

**Reply by kostoskistefan** [reply]
First of all, huge thanks to both of you for actively answering all of my questions. I really appreciate it!

> Yes, use MPLAB IPE or similar tool to flash the MCU. If you pic the correct device the address will be defined in the HEX file for flashing the chip. I think I normally burn the chip right in MPLAB IDE when doing development and only go to the IPE when I'm building a batch of boards (10+).

I prefer using the IPE because the IDE gets really slow sometimes, but it's good to know that I can use them both.

>Yes, although you need to create a proper board profile.

I'll try to create a board profile in the upcoming days. Just by skim reading the files, it's not super complicated, it will just take some time to research and configure everything. 

> If you need help, I'm willing to hop on discord or zoom and try to walk you though / refigure out what is needed to be done.

@JacobChrist This is really kind of you. I do not want to waste your precious time, but if I get really stuck and I have no idea how to continue, it will be a pleasure to talk to you and configure everything. In any case, if everything works without problems, it'll be nice to talk to you and review the code just to confirm that everything is good and following your standards.

>It's not only okay, it's actively encouraged. Let us know where it is and we'll feature it on the chipKIT website :)

That's good to know. I don't have that much free time just yet, but once I do, I'll create a tutorial and I'll notify you. Thanks for giving me the opportunity to contribute to the project!

**Reply by JacobChrist** [reply]
@kostoskistefan It's not wasting my time enabling others to help maintain and improve the chipKIT platform.

**Reply by kostoskistefan** [reply]
@JacobChrist In that case, I'd be very grateful if you could walk me through the configuration process and hopefully improving the ChipKIT project even more in the future. I'd also have the chance to ask questions to better understand how everything works. That being said, both Discord and Zoom work for me. Where should I contact you about setting a meeting date and time, since we probably live in different time zones?

**Comment by JacobChrist** [comment]
@kostoskistefan my Discord is: Papa J#6694

Unless I'm asleep, I'm on-line.

**Reply by kostoskistefan** [reply]
I've just sent you a friend request. Looking forward to talking with you!
